lib/bitcoin/store/utxo_db.rb

```diff
@@ -0,0 +1,260 @@
+require 'leveldb'
+
+module Bitcoin
+  module Store
+    class UtxoDB
+
+      KEY_PREFIX = {
+        out_point: 'o',        # key: out_point(tx_hash and index), value: Utxo
+        script: 's',           # key: script_pubkey and out_point(tx_hash and index), value: Utxo
+        height: 'h',           # key: block_height and out_point, value: Utxo
+        tx_hash: 't',          # key: tx_hash of transaction, value: [block_height, tx_index]
+        block: 'b',            # key: block_height and tx_index, value: tx_hash
+        tx_payload: 'p',       # key: tx_hash, value: Tx
+      }
+
+      attr_reader :level_db, :logger
+
+      def initialize(path = "#{Bitcoin.base_dir}/db/utxo", wallet = nil)
+        FileUtils.mkdir_p(path)
+        @level_db = ::LevelDB::DB.new(path)
+        @logger = Bitcoin::Logger.create(:debug)
+        if wallet.nil?
+          @wallet = Bitcoin::Wallet::Base.current_wallet
+        else
+          @wallet = wallet
+        end
+      end
+
+      def close
+        level_db.close
+      end
+
+      def version
+        1
+      end
+
+      # @param  [String] hashes - array of String from hash160
+      #                         - can also be hash of redeem script
+      # @param  [Bitcoin::Tx] tx - incoming tx.
+      # @param  [Integer] block_height - block height number.
+      # @return [Bitcoin::OutPoint ...] 
+      def filter_and_save_tx(hashes, tx, block_height=nil)
+        out = []
+
+        tx.outputs.each_with_index do |output, index|
+          if (output.script_pubkey.p2pkh? or output.script_pubkey.p2wpkh?) and !(output.script_pubkey.addresses & hashes).empty?
+            out.push(save(tx, index, output, block_height))
+
+          elsif (output.script_pubkey.p2sh? or output.script_pubkey.p2wsh?) and hashes.include?(output.script_pubkey.to_s)
+            save(tx, index, output, block_height)
+            out.push(save(tx, index, output, block_height))
```

Comment:
Need to call `save` method twice in this case?
The second call looks with no mean